### DISCUSSION TITLE
Data Alignment

### DISCUSSION THREAD

**Discussion by SeeRich** [hosseinmoein/DataFrame#217] [Ideas]
My current project uses vector<T, Alloc> to get allocator-aware vectors. This is mainly to use these vectors with SIMD operations that require cache-aligned data. For example, my default custom allocator uses 64 byte alignment. I don't see a way that I can use these vectors to initialize a DataFrame because the interfaces all take vector<T> and not vector<T, Alloc>. Would you be open to supporting the allocator template type? I'm not sure what magnitude of change this represents to the library at the moment, just wanted to start the discussion.

**Comment by hosseinmoein** [comment]
This may require some changes in DataFrame but not much . It just makes the interface a little more involved.

But if you have regular stl vectors of long double, wouldn’t you accomplish the same thing

**Reply by SeeRich** [reply]
Ok, was just curious if SIMD was something you planned to use in the future. I don't think a vector of long double would help. The long double is guaranteed to be at least 64 bits but can be greater if there is hardware support. However, the data size doesn't have anything to do with how the data is aligned in memory.

Here is a post that provides the SIMD motivation and even example code of a modern C++ custom allocator!

https://stackoverflow.com/questions/60169819/modern-approach-to-making-stdvector-allocate-aligned-memory

**Comment by hosseinmoein** [comment]
Yeah, I can incorporate that ability. I can write an aligned allocator and then switch all vectors to that optionally. It doesn't have to be across all columns
 I just need to find time to do it

**Comment by hosseinmoein** [comment]
@RichProgrammerEE ,

If you want you can watch this PR and comment on it to implement aligned memory allocation
https://github.com/hosseinmoein/DataFrame/pull/218

It it going to take me a little time to implement this all the way through

**Comment by SeeRich** [comment]
Oh wow, you have made some quick progress. It looks great, thank you!

**Comment by hosseinmoein** [comment]
This is now done and merged to master

**Reply by SeeRich** [reply]
Thank you! I will test it out later this week.
